pallets/gated-marketplace/src/functions.rs

```diff
@@ -150,9 +151,328 @@ impl<T: Config> Pallet<T> {
+        let owner_item = pallet_uniques::Pallet::<T>::owner(collection_id, item_id).ok_or(Error::<T>::OwnerNotFound)?;
+
+        //ensure only owner of the item can call the extrinic
+        ensure!(owner_item == authority.clone(), Error::<T>::NotOwner);
+
+        // Get the account_id of the offer creator (the buyer)
+        let buy_offer_creator = Self::get_offer_creator(offer_id).map_err(|_| Error::<T>::OfferNotFound)?;
+
+        //ensure owner is not the same as the buy_offer_creator
+        ensure!(owner_item != buy_offer_creator.clone(), Error::<T>::CannotTakeOffer);
+
+        //ensure the selected item has a valid offer_id in OffersInfo
+        // This offer_id corresponds to the buy offer
+        ensure!(<OffersInfo<T>>::contains_key(offer_id), Error::<T>::OfferNotFound);
+
+        //ensure the offer_id exists in OffersByItem
+        Self::does_exist_offer_id_for_this_item(collection_id, item_id, offer_id)?;
+
+        //TODO: ensure the offer is not expired
+
+        //ensure the offer is open and available
+        ensure!(Self::is_offer_status(offer_id, OfferStatus::Open), Error::<T>::OfferIsNotAvailable);
+
+        //Get the offered price
+        let offerred_price =  Self::get_offer_price(offer_id).map_err(|_| Error::<T>::OfferNotFound)?;
+        let total_buy_offer_creator = T::LocalCurrency::total_balance(&buy_offer_creator);
+        //ensure the buy_offer_creator has enough balance to buy the item
+        ensure!(total_buy_offer_creator >= offerred_price, Error::<T>::NotEnoughBalance);
+        //Transfer the balance to the owner of the item
+        T::LocalCurrency::transfer(&buy_offer_creator, &owner_item, offerred_price, KeepAlive)?;
+
+        //Use uniques transfer function to transfer the item to the market_participant
+        pallet_uniques::Pallet::<T>::do_transfer(collection_id, item_id, buy_offer_creator.clone(), |_, _|{
+            Ok(())
+        })?;
+
+        //update offer status from all marketplaces
+        Self::update_offers_status(buy_offer_creator.clone(), collection_id, item_id, marketplace_id)?;
+
+        //remove all the offers associated with the item
+        Self::delete_all_offers_for_this_item(collection_id, item_id )?;
+
+        Self::deposit_event(Event::OfferWasAccepted(offer_id, buy_offer_creator));
+        Ok(())
+    }
+
+    pub fn do_duplicate_offer(authority: T::AccountId, offer_id: [u8;32], marketplace_id: [u8;32], collection_id: T::CollectionId, item_id: T::ItemId, modified_price: BalanceOf<T>) -> DispatchResult{
+        //ensure new marketplace_id exits
+        ensure!(<Marketplaces<T>>::contains_key(marketplace_id), Error::<T>::MarketplaceNotFound);
+
+        //ensure that the offer_id exists
+        ensure!(<OffersInfo<T>>::contains_key(offer_id), Error::<T>::OfferNotFound);
+
+        //ensure the offer_id exists in OffersByItem
+        Self::does_exist_offer_id_for_this_item(collection_id, item_id, offer_id)?;
+
```

Comment:
Shouldn't we check the offer status is open?

Reply:
Yes we should, but I think we don't need this extrinsic anymore. (See previous comment)